src/Corax/Queries/BinaryMatch.Boolean.cs

```diff
@@ -111,88 +168,109 @@ static int FillFunc(ref BinaryMatch<TInner, TOuter> match, Span<long> matches)
-                long* outerMatchesPtr = (long*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(outerMatches));
-                long* matchesStartPtr = (long*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(matches));
+                var innerSlice = buffer.Slice(0, idx);
+                int innerCount = inner.Fill(innerSlice);
+                var outerSlice = buffer.Slice(idx, idx);
+                int outerCount = outer.Fill(outerSlice);
+                Debug.Assert(innerSlice.Length == outerSlice.Length);
 
-                if (innerCount == 0)
+                // If we depleted everything we are done. 
+                if (innerCount == 0 && outerCount == 0)
                 {
-                    Unsafe.CopyBlockUnaligned(matchesStartPtr, outerMatchesPtr, (uint)outerCount * sizeof(long));
-                    return outerCount;
-                }
-                if (outerCount == 0)
-                {
-                    Unsafe.CopyBlockUnaligned(matchesStartPtr, innerMatchesPtr, (uint)innerCount * sizeof(long));
-                    return innerCount;
+                    idx = 0;
+                    goto END; // Nothing more to do here, we are done. 
                 }
 
-                var innerMatchesPtrEnd = innerMatchesPtr + innerCount;
-                var outerMatchesPtrEnd = outerMatchesPtr + outerCount;
+                // We need to know if we have to run another who is gonna come next. The heuristic is to take the one with the smaller
+                // index, so we can ensure to get as many repeated documents as possible in the batch. The rationale of why we do this
+                // is to avoid having to do repeated work higher up in the query. 
 
-                long* matchesPtr = matchesStartPtr;
-                long* matchesPtrEnd = matchesStartPtr + matches.Length;
+                long innerMax = innerCount == 0 ? long.MaxValue : innerSlice[innerCount - 1];
+                long outerMax = outerCount == 0 ? long.MaxValue : outerSlice[outerCount - 1];
 
-                while (innerMatchesPtr < innerMatchesPtrEnd && outerMatchesPtr < outerMatchesPtrEnd)
+                bool isOuterNext;
+                if (innerCount == 0)
+                    isOuterNext = true;
+                else if (outerCount == 0)
+                    isOuterNext = false;
+                else
+                    isOuterNext = innerMax > outerMax;               
+
+                // This is the first run, merge on external buffer and copy back.
+                idx = MergeHelper.Or(matches, innerSlice.Slice(0, innerCount), outerSlice.Slice(0, outerCount));                
+                if (idx == matches.Length)
+                    return idx; // The buffer is full, we are done. 
+
+                // Most of the times we may not need to execute this part, specially when 
+                var leftoverMatches = buffer.Slice(idx);
+                while (leftoverMatches.Length > (totalLength / 32))
```

Comment:
What happens when we do `limit 2` ? For example.

Reply:
Raven will create at least 4K buffer (size depends on query context), even if you wrote `limit 2`.

Reply:
It will execute until it fills it. Anything below 32 will execute until it fills, no matter what. This only trigger when the numbers are big.

Reply:
Most likely that behavior will change when we tackle https://issues.hibernatingrhinos.com/issue/RavenDB-17757 . But for now that is how it is done.